ocaml/src/repair.ml

```diff
@@ -0,0 +1,216 @@
+(*
+Copyright 2015 Open vStorage NV
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*)
+
+open Prelude
+open Lwt_bytes2
+open Lwt.Infix
+
+let repair_object_generic
+      (alba_client : Alba_base_client.client)
+      ~namespace_id
+      ~manifest
+      ~problem_osds
+      ~problem_fragments
+  =
+  let open Nsm_model in
+  Lwt_log.debug_f "_repair_object_generic ~namespace_id:%li ~manifest:%s ~problem_osds:%s ~problem_fragments:%s"
+                  namespace_id (Manifest.show manifest)
+                  ([%show : int32 list] (Int32Set.elements problem_osds))
+                  ([%show : (int * int) list] problem_fragments)
+  >>= fun () ->
+
+  let object_name = manifest.Manifest.name in
+  let object_id = manifest.Manifest.object_id in
+
+  let locations = manifest.Manifest.fragment_locations in
+  let fragment_checksums = manifest.Manifest.fragment_checksums in
+  let fragment_info =
+    Layout.combine
+      locations
+      fragment_checksums
+  in
+
+  alba_client # get_ns_preset_info ~namespace_id >>= fun preset ->
+  alba_client # nsm_host_access # get_gc_epoch ~namespace_id >>= fun gc_epoch ->
+
+  let fragment_checksum_algo =
+    preset.Albamgr_protocol.Protocol.Preset.fragment_checksum_algo in
+
+  let es, compression = match manifest.Manifest.storage_scheme with
+    | Storage_scheme.EncodeCompressEncrypt (es, c) -> es, c in
+  let enc = manifest.Manifest.encrypt_info in
+  let encryption = Albamgr_protocol.Protocol.Preset.get_encryption preset enc in
+  let decompress = Fragment_helper.maybe_decompress compression in
+  let k, m, w = match es with
+    | Encoding_scheme.RSVM (k, m, w) -> k, m, w in
+  let w' = Encoding_scheme.w_as_int w in
+
+  let version_id = manifest.Manifest.version_id + 1 in
+
+  Lwt_list.map_s
+    (fun (chunk_id, chunk_location) ->
+
+     alba_client # get_namespace_osds_info_cache ~namespace_id >>= fun osds_info_cache' ->
+
+     let _, ok_fragments, fragments_to_be_repaired =
+       List.fold_left
```

Comment:
The 2 resulting lists are of a different type though, so would require an extra transformation step after the partition... I'll leave it as is